### ISSUE TITLE
[Error] `/etc/localtime` failed to bind mount during distrobox-init

### ISSUE THREAD

**Issue by weiren2** [89luca89/distrobox#1029]
**Describe the bug**
The command `distrobx enter ...` failed due to bind mount failure of `/etc/localtime`. The backing filesystem is ZFS (2.2.0) with native overlay enabled.

**To Reproduce**
- Create a new distrobox: `distrobox create --name ubuntu-22.04 --image ubuntu:22.04 --home /path/to/ubuntu-home`
- Enter: `distrobox enter ubuntu-22.04`

**Expected behavior**
We should expect the commands finish without error, initialize all the mount points successfully, and enter the distrobox/container.

**Logs**
- Log of `distrobox enter ubuntu-22.04`: [distrobox-enter.log](https://github.com/89luca89/distrobox/files/13175765/distrobox-enter.log)
- Log from `podman`: [podman.log](https://github.com/89luca89/distrobox/files/13175825/podman.log)

**Desktop (please complete the following information):**
 - Are you using podman, docker or lilipod? 
   - **podman**
 - Which version or podman, docker or lilipod?
   - **podman 4.7.1**
 - Which version of distrobox?
   - **distrobox 1.5.0.2**
 - Which host distribution?
   - **Arch Linux**
 - How did you install distrobox?
   - **From Arch's repo (extra)**

**Additional context**
This issue seems to happen only when the backing filesystem is ZFS. I did not encounter such error when using xfs or btrfs.

As a workaround, I moved `/etc/localtime` from `HOST_MOUNTS_RO` to `HOST_MOUNTS`, which I think avoids being (bind) mounted as read-only. I'm not sure if this is a valid fix, but it shouldn't break any permission work assuming rootless podman/docker. https://github.com/89luca89/distrobox/blob/9309e1ebe5c70fd1f9b5286fd3a76f3e7bd995bd/distrobox-init#L1394-L1441
And it seemed working (at least the commands finished without errors/warnings and I can continue working inside container).

Let me know if you need any further testing.

**Comment by rrahl0** [comment]
This bug is fixed in the main version. So you would need to wait for a new release, or the git version.

**Comment by weiren2** [comment]
> This bug is fixed in the main version. So you would need to wait for a new release, or the git version.

The same error still occurred when I tested with the latest git version just now.

@rrahl0 Which bug or commit are you referring to (that is supposed to fix the issue)? I can help look into the issue to see if the fix can be applied.

**Comment by rrahl0** [comment]
@weiren2 pretty sure it was https://github.com/89luca89/distrobox/commit/24ff0c0cacb960a6917d663f6779f2bfe342e678. at least on my fedora/tumbleweed installs it works fine

**Comment by weiren2** [comment]
The error I'm getting is: (also in the complete log)
```bash
+ mount --rbind -o ro /run/host/usr/share/zoneinfo/America/Chicago /etc/localtime
mount: /etc/localtime: filesystem was mounted, but any subsequent operation failed: Unknown error 5005.
+ printf Warning: failed to bind mount %s to %s\n /run/host/usr/share/zoneinfo/America/Chicago /etc/localtime
+ return 1
+ [ 1 -ne 0 ]
+ printf Error: An error occurred\n
```
My guess is that it may be related to zfs being used as the underlying filesystem, which includes additional mount option like `nodev` (and it cannot be removed with rootless). By moving `/etc/localtime` from the read-only mount points ([`HOST_MOUNTS_RO`](https://github.com/89luca89/distrobox/blob/bf65e627b7b5c9b505932a671618c378aa84c48f/distrobox-init#L1395)) to the read-write mount points ([`HOST_MOUNTS`](https://github.com/89luca89/distrobox/blob/bf65e627b7b5c9b505932a671618c378aa84c48f/distrobox-init#L1421C1-L1421C12)), I was able to finishing running the distrobox-init script without errors on my particular system.

@89luca89 Is there a specific reason that `/etc/localtime` needs to be mounted as read-only?
FYI I found this PR (https://github.com/containers/toolbox/pull/1340) in the toolbox project that fixed a similar issue.

**Comment by weiren2** [comment]
I did some further debugging. It turns out that the issue is related to missing mount flags when bind mounting as `ro`. This function [`get_locked_mount_flags`](https://github.com/89luca89/distrobox/blob/bf65e627b7b5c9b505932a671618c378aa84c48f/distrobox-init#L211) does not check for mount flags if the mount point is a file, but `/etc/localtime` is indeed a file (and it can have flags like `nodev`).

I've created a PR #1041 to address this issue.

This issue does not happen when using BTRFS or XFS as the underlying root file system, since there is no additional flags like `nodev` when using kernel-native filesystems for root. But with ZFS as root file system, the mount point ends up having `nodev` by default.

On a side note, the workaround (changing to read-write mount) is just a red herring. Changing the mount mode to `rw` works because there is no remount.  [Source: [containerd/containerd/pull/6913](https://github.com/containerd/containerd/pull/6913#discussion_r868761490).]
